### ISSUE TITLE
How to render an ipython note book using nbconvert in the course page?

### ISSUE THREAD

**Issue by dzhuang** [inducer/relate#343]
According to [nbconvert docs](https://nbconvert.readthedocs.io/en/latest/nbconvert_library.html), the note book can be rendered like [this](http://jakevdp.github.io/blog/2013/04/15/code-golf-in-python-sudoku/), I am wondering if we can do that in RELATE course pages. Any suggestions?

**Comment by dzhuang** [comment]
I think the convert snippet should be put before [this line](https://github.com/inducer/relate/blob/master/course/content.py#L536), am I right?

**Comment by inducer** [comment]
That's a cute capability--I like it. I think the right way to do that would be as function stuck into Jinja's global namespace, rather than additional custom syntax.

http://jinja.pocoo.org/docs/2.9/api/#global-namespace

Something like

```
{{ render_notebook_cells("abc/xyz.ipynb", [12, 13, 15]) }}
```

(Not totally sure about Jinja syntax--but possibly similar. :)

**Comment by dzhuang** [comment]
Thanks, I'll try to submit something to do that.

**Comment by dzhuang** [comment]
Then that global namespace should be putted right before  [this line](https://github.com/inducer/relate/blob/master/course/content.py#L536), because the ``repo`` (which is supposed to locate the file) won't be used in sucessive expanding method, am I right?

**Comment by inducer** [comment]
No. Since you're likely looking to expand those notebook cells inside of existing markup, [this](https://github.com/inducer/relate/blob/master/course/content.py#L880) is the expansion you're looking for. Just stick a function to generate that text into `jinja_env` (without modifying the passed value, obviously).

xref: http://jinja.pocoo.org/docs/2.9/api/#jinja2.Template.render

**Comment by dzhuang** [comment]
Then that should be

```
{% render_notebook_cells("abc/xyz.ipynb", [12, 13, 15]) %}
```
instead of 
```
{{ render_notebook_cells("abc/xyz.ipynb", [12, 13, 15]) }}
```
With the later, it raises ``UndefinedError`` for ``render_notebook_cells`` in ``expand_yaml_macros``, when the yaml is first rendered, before expanding the jinja macros. If that is True, we will have to define a macro and use the global namespace like what I've done in #115 .

**Comment by dzhuang** [comment]
A simple test (in sandbox):

```
type: HumanGradedTextQuestion
id: hgtext
value: 5
widget: textarea
prompt: |

  # Mark Twain
  Discuss the literary work of Mark Twain.

  {{ test() }}

rubric: |

    Has something interesting to say.
```

If we define ``test()`` before [#line879](https://github.com/inducer/relate/blob/master/course/content.py#L879) with:
```
def test():
    return "test"

env.globals["test"] = test
```
it raises ``UndefinedError:  'test' is undefined.

If we put that before [#line536](https://github.com/inducer/relate/blob/master/course/content.py#L536),  i.e., 
```
def test():
    return "test"

jinja_env.globals["test"] = test
```
it worked as expected.

**Comment by inducer** [comment]
What you're saying overall is that the `{{ test() }}` invocation gets caught by the YAML-level Jinja expansion and not the markup-level Jinja expansion. That doesn't make a whole lot of sense to me. The content of the block literal (after the `prompt: |` line) is supposed to be wrapped in `{% raw %}/{% endraw %}`, which should prevent expansion in YAML-level Jinja. Could you verify that that's the case by printing the generated template? Do you maybe have trailing spaces after the `|` character?

> Maybe the work around can be something like this:
> 
> ```
> [IPYNB]render_notebook_cells("abc/xyz.ipynb", [12, 13, 15])[/IPYNB]
> ```

Definitely not. Relate's YAML syntax is already cluttered (and intransparent) enough, with a fair number of unexpected interactions that had to be mitigated. There's no way I'm adding any more special syntax.

Also, for what issue does this provide a workaround?

> ```
> env.globals["test"] = test
> ```

No. Just construct the right expansion context from the get-go. In-place modification is a big no-no.

**Comment by dzhuang** [comment]
> The content of the block literal (after the prompt: | line) is supposed to be wrapped in {% raw %}/{% endraw %}, which should prevent expansion in YAML-level Jinja. Could you verify that that's the case by printing the generated template? Do you maybe have trailing spaces after the | character?

That's true for flow pages, but not the case in sandbox. Is that the reason I got ``UndefinedError``? I am debugging in sandbox.

**Comment by inducer** [comment]
Thanks for working on #344. I imagine that addresses your question above. 👍

**Comment by dzhuang** [comment]
Your suggestions helped me to avoid heading for wrong directions, thanks your patience.

**Comment by dzhuang** [comment]
Closed via 5451d60aa8710420de495d0248e4b0c4d53c7fc1.
